what4/src/What4/Utils/BVDomain.hs

```diff
@@ -1,18 +1,503 @@
+bvdMask x =
+  case x of
+    BVDAny mask -> mask
+    BVDInterval mask _ _ -> mask
+
+--------------------------------------------------------------------------------
+
+-- | @halfRange n@ returns @2^(n-1)@.
+halfRange :: (1 <= w) => NatRepr w -> Integer
+halfRange w = bit (widthVal w - 1)
+
+--------------------------------------------------------------------------------
+-- Projection functions
+
+-- | Convert domain to list of ranges.
+ranges :: NatRepr w -> BVDomain w -> [(Integer, Integer)]
+ranges _w x =
+  case x of
+    BVDAny mask -> [(0, mask)]
+    BVDInterval mask xl xd
+      | xh > mask -> [(0, xh .&. mask), (xl, mask)]
+      | otherwise -> [(xl, xh)]
+      where xh = xl + xd
+
+-- | Return value if this is a singleton.
+asSingleton :: BVDomain w -> Maybe Integer
+asSingleton x =
+  case x of
+    BVDAny _ -> Nothing
+    BVDInterval _ xl xd
+      | xd == 0 -> Just xl
+      | otherwise -> Nothing
+
+isSingletonZero :: BVDomain w -> Bool
+isSingletonZero x =
+  case x of
+    BVDInterval _ 0 0 -> True
+    _ -> False
+
+isBVDAny :: BVDomain w -> Bool
+isBVDAny x =
+  case x of
+    BVDAny {} -> True
+    BVDInterval {} -> False
+
+-- | Return unsigned bounds for domain.
+ubounds :: BVDomain w -> (Integer, Integer)
+ubounds a =
+  case a of
+    BVDAny mask -> (0, mask)
+    BVDInterval mask al aw
+      | ah > mask -> (0, mask)
+      | otherwise -> (al, ah)
+      where ah = al + aw
+
+-- | Return signed bounds for domain.
+sbounds :: (1 <= w) => NatRepr w -> BVDomain w -> (Integer, Integer)
+sbounds w a = (lo - delta, hi - delta)
+  where
+    delta = halfRange w
+    (lo, hi) = ubounds (add a (BVDInterval (bvdMask a) delta 0))
+
+-- | Return true if domains contain a common element.
+domainsOverlap :: BVDomain w -> BVDomain w -> Bool
+domainsOverlap a b =
+  case a of
+    BVDAny _ -> True
+    BVDInterval _ al aw ->
+      case b of
+        BVDAny _ -> True
+        BVDInterval mask bl bw ->
+          diff <= bw || diff + aw > mask
+          where diff = (al - bl) .&. mask
+
+eq :: BVDomain w -> BVDomain w -> Maybe Bool
+eq a b
+  | Just x <- asSingleton a
+  , Just y <- asSingleton b = Just (x == y)
+  | domainsOverlap a b == False = Just False
+  | otherwise = Nothing
+
+-- | Check if all elements in one domain are less than all elements in other.
+slt :: (1 <= w) => NatRepr w -> BVDomain w -> BVDomain w -> Maybe Bool
+slt w a b
+  | a_max < b_min = Just True
+  | a_min >= b_max = Just False
+  | otherwise = Nothing
+  where
+    (a_min, a_max) = sbounds w a
+    (b_min, b_max) = sbounds w b
+
+-- | Check if all elements in one domain are less than all elements in other.
+ult :: (1 <= w) => BVDomain w -> BVDomain w -> Maybe Bool
+ult a b
+  | a_max < b_min = Just True
+  | a_min >= b_max = Just False
+  | otherwise = Nothing
+  where
+    (a_min, a_max) = ubounds a
+    (b_min, b_max) = ubounds b
+
+-- | Return @Just@ if every bitvector in the domain has the same bit
+-- at the given index.
+testBit ::
```

Comment:
Is this easier to understand if we state it in terms of `unknowns`?

Reply:
Maybe, but `unknowns` is more expensive to compute (O(lg n) recursive calls)